Add additional configuration refresh stats

samsymons opened the pull request:
Task/Issue URL: https://app.asana.com/0/1193060753475688/1198516960588141
CC: @bwaresiak 

**Description**:

This PR adds extra stats to the configuration refresh process, to better understand where errors may be occurring.

1. An early return has been added when scheduling a background task on a device which doesn't have permission, to avoid sending needless error pixels
1. A new stat has been added which is tracked when a background task is expired by the device
1. Stats have been added on a per-config file basis to help us understand if any files are causing problems

**Steps to test this PR**:
1. Manually trigger a configuration refresh and check the parameters being sent with the pixel call
1. Schedule a background task and manually force it to expire (instructions can be found via a comment in the code) to check that the expiration stat is tracked

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [x] iPad

**OS Testing**:

* [ ] iOS 11
* [ ] iOS 12
* [ ] iOS 13
* [x] iOS 14

samsymons committed:
Prevent scheduling a task if the device doesn’t have permission.

This is happening in the AppDelegate instead of the `scheduleBackgroundRefreshTask` function because of UIApplication only being usable from the main thread.

samsymons committed:
Track when tasks are expired by the OS.

samsymons committed:
Add a comment about task expiration.

samsymons committed:
Track statistics on the config refresh process.

This will help to determine if any are slow and causing problems with the config refresh process.

bwaresiak commented on Core/ContentBlockerLoader.swift:
```diff
@@ -109,11 +121,17 @@ public class ContentBlockerLoader {
             
```
I wonder: maybe we also should mark `httpsBloomFilter` as completed if in the above block, as effectively it won't be invoked (and should not be marked as an potential issue).

samsymons replied:
Great catch, thanks!

bwaresiak commented on DuckDuckGo/AppUserDefaults.swift:
```diff
@@ -202,4 +211,67 @@ extension AppUserDefaults: AppConfigurationFetchStatistics {
+        set {
+            userDefaults?.setValue(newValue, forKey: Keys.backgroundFetchTaskExpirationCount)
+        }
+    }
```
Can we use property wrappers here? Wouldn't that simplify the code?

samsymons replied:
Great call, we should do that. I added it this way for consistency with the current approach, but new additions should definitely be using the more modern approach – I'll update this, thanks!

samsymons committed:
Swap to property wrappers for config fetch stats.

samsymons committed:
Add a call to the HTTPS Bloom filter progress block when returning early.

This makes sure that we are marking the Bloom filter as downloaded when we are deliberately not downloading it, otherwise it might get mistaken with an attempted download that has gone wrong.

samsymons committed:
Rename a typealias.

bwaresiak reviewed (approved):
LGTM! 👍
